signer-service/src/api/services/siwe.service.js

```diff
@@ -0,0 +1,140 @@
+    address: address,
+    message: siweMessage.toMessage(), // Validation must be done on the message as string
+    signature,
+  });
+
+  if (!valid) {
+    throw new ValidationError('Invalid signature');
+  }
+
+  // Perform additional checks to ensure message fields
+  if (siweMessage.nonce !== nonce) {
+    throw new ValidationError('Nonce mismatch');
+  }
+
+  if (siweMessage.expirationTime && new Date(siweMessage.expirationTime) < new Date()) {
+    throw new ValidationError('Message has expired');
+  }
+
+  return siweMessage;
+};
+
+// Since the message is created in the UI, we need to verify the fields of the message
+const verifyInitialMessageFields = (siweMessage) => {
+  // Fields we validate on initial
+  const domain = siweMessage.domain;
+  const uri = siweMessage.uri;
+  const scheme = siweMessage.scheme; // must be https
+  const chainId = siweMessage.chainId;
+  const expirationTime = siweMessage.expirationTime;
+
+  if (!VALID_SIWE_LOGIN_ORIGINS.includes(uri)) {
+    throw new ValidationError('Origin not in the list of allowed origins');
+  }
+
+  if (!VALID_SIWE_DOMAINS.includes(domain)) {
+    throw new ValidationError('Incorrect domain');
+  }
+
+  if (!VALID_SIWE_CHAINS.includes(chainId)) {
+    throw new ValidationError('Incorrect chain ID');
+  }
+
+  if (scheme !== 'https') {
+    throw new ValidationError('Scheme must be https');
+  }
+
+  if (!expirationTime || isNaN(new Date(expirationTime).getTime())) {
+    throw new ValidationError('Must define a valid expiration time');
+  }
+
+  // Check if expiration is within a reasonable range from current time
+  const currentTime = new Date().getTime();
+  const expirationTimestamp = new Date(expirationTime).getTime();
+
+  const expirationGracePeriod = 1000 * 60; // 1 minute
+  const expirationPeriodMs = DEFAULT_LOGIN_EXPIRATION_TIME_HOURS * 60 * 60 * 1000;
+  const expectedMinExpirationTimestamp = currentTime + expirationPeriodMs - expirationGracePeriod;
+  const expectedMaxExpirationTimestamp = currentTime + expirationPeriodMs;
+
+  if (expirationTimestamp < expectedMinExpirationTimestamp) {
+    throw new ValidationError('Expiration time is too low');
+  }
+
+  if (expirationTimestamp > expectedMaxExpirationTimestamp) {
+    throw new ValidationError('Expiration time is too high');
+  }
+};
+
+const verifyAndStoreSiweMessage = async (nonce, signature, siweMessage) => {
+  const validatedMessage = await verifySiweMessage(nonce, signature, siweMessage);
+
+  // Perform additional checks to ensure message fields are valid
+  verifyInitialMessageFields(validatedMessage);
+
+  // Verification complete. Update the map and append the message.
+  const siweData = siweMessagesMap.get(nonce);
+  siweData.siweMessage = validatedMessage;
+  siweMessagesMap.set(nonce, siweData);
+
+  // Remove messages with same address from map except for the one we just verified.
+  // This will keep one valid session per address.
+  siweMessagesMap.forEach((data, nonce) => {
```

Comment:
The original reasoning was to keep only one "session" per user, per address. This and the logout (that we decided not to do), where the original rationale for adding the map in the first place. Otherwise we would not really need it, it wouldn't add extra security.

Since it is relatively small amounts of logic, I decided to leave this functionality. Totally optional of course.